### PR TITLE
feat(browser): refactor the WebSocket reconnection to assign the peer as id

### PR DESCRIPTION
**Pull request by gdorsi**
To achieve #307 we need a stable identifier for the WebSocket peer across the page sessions.

To achieve that goal I've refactored the WebSocket reconnection flow and added an "onClose" handler option to the websocket peer.

### CODE REVIEW

File: packages/cojson-transport-ws/src/index.ts [hunk 1]
```diff
@@ -18,87 +18,54 @@ export type CreateWebSocketPeerOpts = {
...
-                    .catch((e) =>
-                        console.error("Error while pushing ping timeout", e),
-                    );
+                callback();
             }, 10_000);
+        },
+        clear() {
+            pingTimeout && clearTimeout(pingTimeout);
         }
+    };
```
Comment (gdorsi) [inline packages/cojson-transport-ws/src/index.ts:44]: Extracted the ping timeout management, to simplify the cleanup management

File: packages/cojson-transport-ws/src/index.ts [hunk 2]
```diff
@@ -18,87 +18,54 @@ export type CreateWebSocketPeerOpts = {
...
+    return new Promise<void>((resolve) => {
         if (websocket.readyState === 1) {
             resolve();
         } else {
             websocket.addEventListener("open", resolve, { once: true });
         }
     });
+}
 
+function createOutgoingMessagesManager(websocket: AnyWebSocket, batchingByDefault: boolean) {
```
Comment (gdorsi) [inline packages/cojson-transport-ws/src/index.ts:57]: Abstracted the outgoing messages sending to simplify the batching toggle management

File: packages/cojson-transport-ws/src/index.ts [hunk 3]
```diff
@@ -114,23 +81,114 @@ export function createWebSocketPeer({
...
                 console.log("Trying to close", id, websocket.readyState);
-                if (supportsBatching) {
-                    outgoingMessages.close();
-                }
+                outgoingMessages.close();
+
+                websocket.removeEventListener("message", handleIncomingMsg);
+                websocket.removeEventListener("close", handleClose);
+                pingTimeout.clear();
+                emitClosedEvent();
```
Comment (gdorsi) [inline packages/cojson-transport-ws/src/index.ts:191]: Refactored the close management to detach all the events and clear all the pending timeouts

File: packages/jazz-browser/src/createWebSocketPeerWithReconnection.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
+import { Peer } from "cojson";
+import { createWebSocketPeer } from "cojson-transport-ws";
+
+export function createWebSocketPeerWithReconnection(peer: string, reconnectionTimeout: number | undefined, addPeer: (peer: Peer) => void) {
```
Comment (gdorsi) [inline packages/jazz-browser/src/createWebSocketPeerWithReconnection.ts:4]: Extracted the WebSocket reconnection flow in his own module

File: packages/jazz-browser/src/createWebSocketPeerWithReconnection.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    const firstWsPeer = createWebSocketPeer({
+        websocket: new WebSocket(peer),
+        id: peer,
```
Comment (gdorsi) [inline packages/jazz-browser/src/createWebSocketPeerWithReconnection.ts:7]: Changed the id from `options.peer + "@" + new Date().toISOString()` to have a stable identifier

File: packages/jazz-browser/src/createWebSocketPeerWithReconnection.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    let currentReconnectionTimeout = reconnectionTimeout || 500;
+
+    function onOnline() {
+        console.log("Online, resetting reconnection timeout");
+        currentReconnectionTimeout = reconnectionTimeout || 500;
+    }
+
+    window.addEventListener("online", onOnline);
+
+    async function reconnectWebSocket() {
```
Comment (gdorsi) [inline packages/jazz-browser/src/createWebSocketPeerWithReconnection.ts:22]: Insted of a timeout loop here we rely on the new onClose handler

File: packages/jazz-browser/src/createWebSocketPeerWithReconnection.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    return {
+        peer: firstWsPeer,
+        done: () => {
+            shouldTryToReconnect = false;
+            window.removeEventListener("online", onOnline);
+        }
+    };
+}
+function waitForOnline(timeout: number) {
```
Comment (gdorsi) [inline packages/jazz-browser/src/createWebSocketPeerWithReconnection.ts:57]: Extracted the timeout/online waiting in his own function

File: packages/jazz-browser/src/test/createWebSocketPeerWithReconnection.test.ts
Comment (bensleveritt) [inline packages/jazz-browser/src/test/createWebSocketPeerWithReconnection.test.ts:1]: Folder should be /tests/ I think

**Review by aeplay — APPROVED** [review]

**Pull request merged**